contributors/design-proposals/federated-statefulsets.md

```diff
@@ -0,0 +1,491 @@
+query too. For example the service CNAME record ```etcd.mynamespace.myfederation.svc.example.com```
+will also result in similar local names as elaborated above, especially because at this point
+the headless service, even though federated, will not get any dns names updated into the
+global (federated) dns server. The results will however be different, if the global identities
+are also requested by the user as elaborated in the next section.
+
+### Instance identity and discovery (global identities)
+
+The above is fine for the case of multiple local quora (one in each cluster), but these
+identities cannot work across the clusters, as the local pod IPs cannot be reached across
+the clusters.
+
+The proposal to make it work across the clusters is to create an additional service of type
+&#39;LoadBalancer&#39; for each stateful pod instance that is created in the clusters. The
+name of each service instance will be same as the stateful instance name assigned to the
+stateful replicas. This will be automatically created by the federated statefulset controller,
+on demand, proposed to be controlled by an annotation on the statefulset. If the annotation
+specifies so (users choice), the federation controller will create the additional service of
+type &#39;LoadBalancer&#39; against each pod replica. The key name for the annotation will be:
+```federation.kubernetes.io/statefulset-lb-needed```
+The values could be ```true``` or ```false```. The default value will be ```false```, considering
+that the external load balancer is an expensive resource.
+
+The service name for the LB service, one each against the stateful pod will be the same name as
+the stateful pod unique name. Also note that these particular LB services, although federated,
+will ideally resolve to 1 IP per service. The LB services will be created with proper label
+selectors to ensure the network traffic can reach the designated pods. This would also mean
+that each stateful pod will need to be updated with an additional unique label selector which
+can be matched with by the individual LB services. The suggestion is to use
+```federated-statefulset: <stateful replica unique name>``` as the label selector.
+
+The details of creation and handling the LB services are elaborated in section **Handling
+additional LB Services** later in this doc.
+
+The trick for individual pods being able to discover this additional dns identity through the LB
+service is the federated controller first creating SRV records for each stateful pod in the
+federation dns server against the governing service domain
+```store.mynamespace.myfederation.svc.example.com``` and then later updating an A record of the IP
+resolved for the specific LB service (for example ```store-c1-0.mynamespace.myfederation.svc.example.com```)
+against each of these SRV records.
+A peer pod in a stateful set then can discover its federated peers by doing a dns query against
```

Comment:
I think even in the case of in-cluster statefulsets the sequence of creation does not guarantee that the dns name for the previous stateful replica will necessarily resolve to its ip when the next in sequence replica is spun off.
I think its eventual name resolution, and most of the apps ideally are resilient if the dns name does not resolve to its actual ip in the first hit. In case of federated stateful app, I am not sure if this can really be a problem. As stated in the doc earlier, its proposed already to relax the requirement of the guaranteeing the creation sequence in case of federated statefulsets.